### PR TITLE
Updated err message when retrieving resources by name with --all-name…

### LINKED ISSUE #15834: Feel a little confused to use --all-namespaces flag
The  desciption of --all-namespaces is listing the requested object(s) across all namespaces. Namespace in current context is ignored even if specified with --namespace."

But when I used the following command: 
```#./cluster/kubectl.sh get pods counter  --all-namespaces```
error: namespace may not be empty when retrieving a resource by name
```#./cluster/kubectl.sh get -f ./examples/blog-logging/counter-pod.yaml --all-namespaces```
no namespace set on resource pods "counter"

The error messages are unclear, and the second message isn't labeled as an error even.

### PR DESCRIPTION
**Pull request by k82cn**
fixes #15834

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33546)

### CODE REVIEW

**Comment by k82cn** [comment]
Update UT result:

```
Klauss-MacBook-Pro:kubernetes klaus$ ./_output/bin/kubectl --server=http://127.0.0.1:8080 get pods nginx-701339712-x46an --all-namespaces
error: all namespaces or empty namespace is invalid when retrieving a resource by name
```

File: pkg/kubectl/resource/builder.go [hunk 1]
```diff
@@ -655,7 +655,7 @@ func (b *Builder) visitByResource() *Result {
 			selectorNamespace = ""
 		} else {
 			if len(b.namespace) == 0 {
-				return &Result{singular: isSingular, err: fmt.Errorf("namespace may not be empty when retrieving a resource by name")}
+				return &Result{singular: isSingular, err: fmt.Errorf("all namespaces or empty namespace is invalid when retrieving a resource by name")}
```
Comment (liggitt) [inline pkg/kubectl/resource/builder.go]: I'd prefer we keep track of whether AllNamespaces was called so we can give a more specific error message in that case

Comment (k82cn) [inline pkg/kubectl/resource/builder.go]: OK, I'll try it.

**Comment by k82cn** [comment]
Update UT result of latest code:

```
$ ./_output/bin/kubectl get pods aaa --all-namespaces
error: namespaces may not be all when retrieving a resource by name

$ make check WHAT=pkg/kubectl/resource/
Running tests for APIVersion: v1,apps/v1alpha1,authentication.k8s.io/v1beta1,authorization.k8s.io/v1beta1,autoscaling/v1,batch/v1,batch/v2alpha1,certificates.k8s.io/v1alpha1,extensions/v1beta1,imagepolicy.k8s.io/v1alpha1,policy/v1alpha1,rbac.authorization.k8s.io/v1alpha1,storage.k8s.io/v1beta1,federation/v1beta1
+++ [1011 18:30:04] Running tests without code coverage
ok  	k8s.io/kubernetes/pkg/kubectl/resource	0.080s```
```

**Comment by k82cn** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/kubectl/resource/builder.go, line 658 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33546#-KSf40528At0P4cGtjgX-r1-658:-KTnIG6a-F8aUpm_eR_x:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/ea6b74df3eb9b869f19aea72e4ce14e949a9429e/pkg/kubectl/resource/builder.go#L658)):*
<details><summary><i>Previously, k82cn (Klaus Ma) wrote…</i></summary>
> OK, I'll try it.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33546)*

**Comment by k82cn** [comment]
@liggitt , the failed e2e seems not related with this patch, would help to review it?

File: pkg/kubectl/resource/builder.go [hunk 2]
```diff
@@ -644,7 +646,11 @@ func (b *Builder) visitByResource() *Result {
 			selectorNamespace = ""
 		} else {
 			if len(b.namespace) == 0 {
-				return &Result{singular: isSingular, err: fmt.Errorf("namespace may not be empty when retrieving a resource by name")}
+				errMsg := "namespaces may not be empty when retrieving a resource by name"
+				if b.allNamespace {
+					errMsg = "namespaces may not be all when retrieving a resource by name"
```
Comment (liggitt) [inline pkg/kubectl/resource/builder.go]: what about "a resource cannot be retrieved by name across all namespaces"?

File: pkg/kubectl/resource/builder.go [hunk 3]
```diff
@@ -690,7 +696,11 @@ func (b *Builder) visitByName() *Result {
 		selectorNamespace = ""
 	} else {
 		if len(b.namespace) == 0 {
-			return &Result{singular: isSingular, err: fmt.Errorf("namespace may not be empty when retrieving a resource by name")}
+			errMsg := "namespaces may not be empty when retrieving a resource by name"
+			if b.allNamespace {
+				errMsg = "namespaces may not be all when retrieving a resource by name"
```
Comment (liggitt) [inline pkg/kubectl/resource/builder.go]: "a resource cannot be retrieved by name across all namespaces"?

**Review by liggitt — COMMENTED** [review]
nit on the message, could use a test in test-cmd.sh

**Comment by k82cn** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/kubectl/resource/builder.go, line 651 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33546#-KSf40528At0P4cGtjgX-r2-651:-KTrlHkkdI0TKUxnQjEN:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/1002ba505c6c0810855e8f1c1cb6c0d080665719/pkg/kubectl/resource/builder.go#L651)):*
<details><summary><i>Previously, liggitt (Jordan Liggitt) wrote…</i></summary>
> what about "a resource cannot be retrieved by name across all namespaces"?
</details>
Done.

---

*[pkg/kubectl/resource/builder.go, line 701 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33546#-KSf40528At0P4cGtjgX-r2-701:-KTrlItLXvHS6dio9aba:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/1002ba505c6c0810855e8f1c1cb6c0d080665719/pkg/kubectl/resource/builder.go#L701)):*
<details><summary><i>Previously, liggitt (Jordan Liggitt) wrote…</i></summary>
> "a resource cannot be retrieved by name across all namespaces"?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33546)*

**Comment by k82cn** [comment]
The unit/integration failure is handled in https://github.com/kubernetes/kubernetes/pull/34608 which is not related to this PR.

**Comment by liggitt** [comment]
> The unit/integration failure was handled in #34608 which is not related to this PR.

I meant adding (or modifying) a test-cmd.sh test that tries to use --all-namespaces with a named resource and verifies this message

**Comment by k82cn** [comment]
> I meant adding (or modifying) a test-cmd.sh test that tries to use --all-namespaces with a named resource and verifies this message

wow, got your point :). I'll add a test case in `test-cmd.sh`.

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33546#-:-KTtD49ZU9O4_nkbiure:bfnv7sq)*

**Comment by liggitt** [comment]
#33388 flake, @k8s-bot cvm gke e2e test this

File: hack/make-rules/test-cmd.sh [hunk 4]
```diff
@@ -1569,6 +1569,9 @@ __EOF__
   kube::test::get_object_assert 'pods --namespace=other' "{{range.items}}{{$id_field}}:{{end}}" 'valid-pod:'
   # Post-condition: verify shorthand `-n other` has the same results as `--namespace=other`
   kube::test::get_object_assert 'pods -n other' "{{range.items}}{{$id_field}}:{{end}}" 'valid-pod:'
+  # Post-condition: a resource cannot be retrieved by name across all namespaces
+  output_message=$(! kubectl get pod valid-pod --all-namespaces 2>&1)
```
Comment (liggitt) [inline hack/make-rules/test-cmd.sh]: always pass `"${kube_flags[@]}"` as well

see https://github.com/k82cn/kubernetes/blob/30cb370f5a7b9ce1e28499d89d77f2b10d98692c/hack/make-rules/test-cmd.sh#L797-L799 as an example for testing an expected failure and error message

**Comment by k82cn** [comment]
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[hack/make-rules/test-cmd.sh, line 1573 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33546#-KTx7JvqVY6_3WOaePQq-r3-1573:-KU2pqZHB9Gw_A0YHvCw:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/30cb370f5a7b9ce1e28499d89d77f2b10d98692c/hack/make-rules/test-cmd.sh#L1573)):*
<details><summary><i>Previously, liggitt (Jordan Liggitt) wrote…</i></summary>
> always pass `"${kube_flags[@]}"` as well
> 
> 
> 
> see https://github.com/k82cn/kubernetes/blob/30cb370f5a7b9ce1e28499d89d77f2b10d98692c/hack/make-rules/test-cmd.sh#L797-L799 as an example for testing an expected failure and error message
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33546)*

**Comment by liggitt** [comment]
LGTM

**Pull request merged**